cloud/gke.go

```diff
@@ -0,0 +1,1038 @@
+		DbSaveFailed(c, err, cluster2Db.Name)
+		return false, nil
+	}
+
+	banzaiUtils.LogInfo(banzaiConstants.TagCreateCluster, "Save create cluster into database succeeded")
+	SetResponseBodyJson(c, http.StatusCreated, gin.H{
+		JsonKeyStatus:     http.StatusCreated,
+		JsonKeyResourceId: cluster2Db.ID,
+		JsonKeyData:       gkeCluster,
+	})
+	return true, &cluster2Db
+}
+
+func getBanzaiErrorFromError(err error) *banzaiTypes.BanzaiResponse {
+
+	if err == nil {
+		// error is nil
+		return &banzaiTypes.BanzaiResponse{
+			StatusCode: http.StatusInternalServerError,
+		}
+	}
+
+	googleErr, ok := err.(*googleapi.Error)
+	if ok {
+		// error is googleapi error
+		return &banzaiTypes.BanzaiResponse{
+			StatusCode: googleErr.Code,
+			Message:    googleErr.Message,
+		}
+	}
+
+	// default
+	return &banzaiTypes.BanzaiResponse{
+		StatusCode: http.StatusInternalServerError,
+		Message:    err.Error(),
+	}
+}
+
+func generateClusterCreateRequest(cc GKECluster) *gke.CreateClusterRequest {
+	request := gke.CreateClusterRequest{
+		Cluster: &gke.Cluster{},
+	}
+	request.Cluster.Name = cc.Name
+	request.Cluster.Zone = cc.Zone
+	request.Cluster.InitialClusterVersion = cc.MasterVersion
+	request.Cluster.InitialNodeCount = cc.NodeCount
+	request.Cluster.ClusterIpv4Cidr = cc.ClusterIpv4Cidr
+	request.Cluster.Description = cc.Description
+	request.Cluster.EnableKubernetesAlpha = cc.EnableAlphaFeature
+	request.Cluster.AddonsConfig = &gke.AddonsConfig{
+		HttpLoadBalancing:        &gke.HttpLoadBalancing{Disabled: !cc.HTTPLoadBalancing},
+		HorizontalPodAutoscaling: &gke.HorizontalPodAutoscaling{Disabled: !cc.HorizontalPodAutoscaling},
+		KubernetesDashboard:      &gke.KubernetesDashboard{Disabled: !cc.KubernetesDashboard},
+		//	NetworkPolicyConfig:      &gke.NetworkPolicyConfig{Disabled: !cc.NetworkPolicyConfig},
+	}
+	request.Cluster.Network = cc.Network
+	request.Cluster.Subnetwork = cc.SubNetwork
+	request.Cluster.LegacyAbac = &gke.LegacyAbac{
+		Enabled: cc.LegacyAbac,
+	}
+	request.Cluster.MasterAuth = &gke.MasterAuth{
+		Username: "admin",
+	}
+	request.Cluster.NodeConfig = cc.NodeConfig
+	return &request
+}
+
+func GetGoogleServiceClient() (*gke.Service, error) {
+
+	// See https://cloud.google.com/docs/authentication/.
+	// Use GOOGLE_APPLICATION_CREDENTIALS environment variable to specify
+	// a service account key file to authenticate to the API.
+
+	client, err := google.DefaultClient(context.Background(), gke.CloudPlatformScope)
+	if err != nil {
+		banzaiUtils.LogFatalf(banzaiConstants.TagCreateCluster, "Could not get authenticated client: %v", err)
+		return nil, err
+	}
+	service, err := gke.New(client)
+	if err != nil {
+		banzaiUtils.LogFatalf(banzaiConstants.TagCreateCluster, "Could not initialize gke client: %v", err)
+		return nil, err
+	}
+	banzaiUtils.LogInfof(banzaiConstants.TagCreateCluster, "Using service acc: %v", service)
+	return service, nil
+}
+
+// Struct of GKE
+type GKECluster struct {
```

Comment:
Maybe this can be put inside banzai-types.

Reply:
GKECluster is only used inside pipeline. The requests use other structs.